[#44] Simplify editing the CSR

sam-bertin opened the pull request:
Closes #44

🚸 Editing a CSR
Updated the UX to edit a CSR.  Removed the should renew check box. This could be done because a CSR is function of the subject information and the private key. Therefore we only stored the private key when creating the instance and regenerate the CSR on every save (it will stay the same if the subject did not change). Once the CSR is signed (upload a valid certifcate), we display a message to the user to let him know and make the field readonly so the CSR can't be edited.

sam-bertin committed:
[#44] Simplify editing the CSR

🚸 Editing a CSR
Updated the UX to edit a CSR.  Removed the should renew check box. This could be done because a CSR is function of the subject information and the private key. Therefore we only stored the private key when creating the instance and regenerate the CSR on every save (it will stay the same if the subject did not change). Once the CSR is signed (upload a valid certifcate), we display a message to the user to let him know and make the field readonly so the CSR can't be edited.

sergei-maertens commented on simple_certmanager/admin.py:
```diff
@@ -89,16 +88,66 @@ class SigningRequestAdmin(admin.ModelAdmin):
 
+    def get_object(
+        self, request: HttpRequest, object_id: str, from_field: str | None = None
+    ):
+        """
+        Here I only access the object so I can check which message to display
+        to help the user. I don't need to change the object in any way.
+        """
+        object = super().get_object(request, object_id, from_field)
+        if object and object.public_certificate:
+            messages.warning(
+                request,
+                _(
+                    "This request is signed therefore can not be edited."
+                    " You can safely delete it."
+                ),
+            )
+
+        return object
+
+    def get_readonly_fields(
+        self, request: HttpRequest, obj: SigningRequest | None = None
+    ):
+        "Make the CSR field read-only after it has been signed."
+        readonly_fields = [
+            "csr",
+            "public_certificate",
+        ]
+        if obj and obj.public_certificate:
+            readonly_fields.extend(
+                [
+                    "common_name",
+                    "organization_name",
+                    "country_name",
+                    "state_or_province_name",
+                    "locality_name",
+                    "email_address",
+                ]
+            )
+        return readonly_fields
```
I think you can simplify this a lot by overriding the method `has_change_permission` instead.

The `messages` approach may also not be the best since you can trigger the view which will put the message in the session, but it may then be displayed on another page which would be confusing. It's a bit of a race condition, but I've seen cases where messages are displayed on the login screen *after* logging out of the admin.

sam-bertin replied:
OK for the permission.

Do you see another way to notify the users why their permission has changed though ? We looked for ways to override the default displayed message unsuccessfully. 
 
IMHO, it is preferable to have some information that gives notice to the user that he cannot edit an instance anymore rather than no notice but I also agree that the messages might display when not desired.

sergei-maertens replied:
hrm, I thought this mechanism was viable to automatically make the thing readonly (change vs. view permission), but from your comment I have the impression that is not the case?

sam-bertin replied:
Yes the permission mechanism DOES make the instance read-only.

My issue is in the fact that the instance "magically" switches from editable to read-only. 
As developer, we know why it switched but the user might be clueless about why his instance became read-only and that is what I tried to address with the messages/ get_object method.

sergei-maertens replied:
check - I think it's fine to leave it as is. It's the admin environment, which should not be exposed to people that are completely clueless. Keeping in mind the people who we are building this feature for, I think it's clear enough and we can still document this on read the docs if people ask questions.

It's Good Enough.

sam-bertin replied:
allright. Removed the code and added the relevant doc.

sergei-maertens commented on simple_certmanager/admin.py:
```diff
@@ -89,16 +88,66 @@ class SigningRequestAdmin(admin.ModelAdmin):
 
+    def log_change(self, request, obj, message):
+        "Logs that an object has been successfully changed."
+        if obj.public_certificate:
+            message = _("Signing Request processed and deleted.")
+            super().log_deletion(request, obj, message)
```
this confuses me - if the "save" button is clicked, nothing will be automatically deleted, will there?

CharString replied:
This is a remnant from an experiment we ran this morning. It needs to go.

sergei-maertens commented on simple_certmanager/models.py:
```diff
@@ -87,6 +79,16 @@ def save(self, *args, **kwargs):
                 email=self.email_address,
             )
+        else:
+            if not self.public_certificate:
+                self.csr = generate_csr(
+                    key_pem=self.private_key,
+                    common_name=self.common_name,
+                    organization_name=self.organization_name,
+                    locality=self.locality_name,
+                    state_or_province=self.state_or_province_name,
+                    country=self.country_name,
+                )
```
@CharString I'm confused - I thought the CSR would be generated on the fly and not stored in the DB anymore?

CharString replied:
In a different task  #47

sergei-maertens commented:
@sam-bertin if you use closing keywords in the PR description, then the PR will be linked to the issue, e.g.:

> `Closes #44`

CharString commented on simple_certmanager/forms.py:
```diff
@@ -29,46 +28,18 @@ class SigningRequestAdminForm(forms.ModelForm):
-            )
-            instance.private_key = new_private_key
-            instance.csr = new_csr
+            instance.create_certificate(validated_certificate)
```
I like how this has cleaned up

sam-bertin replied:
ok, removed it !

CharString reviewed (changes requested):
The `log_change` code should be removed.
And I would like to see that `has_change_permission` option.

sam-bertin committed:
[#44] 🎨 Block user from changing a signed request

Make use of the has_change_permission method to simplify the restriction/permission to edit/change a SigningRequest. This is simpler than dynamically loading the reading fields. Added relevant doc.
